src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
+					zap.Time("indexTo", indexMax),
+				}...)
+			}
+		}
+
+		b.log.Info("bootstrap from source for namespace ready", prepareLogFields...)
 	}
-	b.log.Info("bootstrapping from source starting", logFields...)
 
 	nowFn := b.opts.ClockOptions().NowFn()
 	begin := nowFn()
 
-	currStatus, currErr = step.runCurrStep(currRanges)
+	b.log.Info("bootstrap from source started", logFields...)
+	currResults, err := b.src.Read(curr)
 
 	logFields = append(logFields, zap.Duration("took", nowFn().Sub(begin)))
-	if currErr != nil {
-		logFields = append(logFields, zap.Error(currErr))
-		b.log.Info("bootstrapping from source completed with error", logFields...)
-	} else {
-		logFields = append(logFields, currStatus.logFields...)
-		b.log.Info("bootstrapping from source completed successfully", logFields...)
+	if err != nil {
+		errorLogFields := append(logFieldsCopy(logFields), zap.Error(err))
+		b.log.Info("bootstrapping from source completed with error", errorLogFields...)
+	}
+	if err != nil {
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	wg.Wait()
-	if err := xerrors.FirstError(currErr, nextErr); err != nil {
-		return err
+	// Determine the unfulfilled and the unattempted ranges to execute next.
+	next := bootstrap.Namespaces{
+		Namespaces: bootstrap.NewNamespacesMap(bootstrap.NamespacesMapOptions{}),
 	}
+	for _, elem := range namespaces.Namespaces.Iter() {
+		id := elem.Key()
+		namespace := elem.Value()
+
+		currResult, ok := currResults.Results.Get(id)
+		if !ok {
+			return bootstrap.NamespaceResults{},
+				fmt.Errorf("namespace result not returned by bootstrapper: %v", id.String())
+		}
+
+		currNamespace, ok := curr.Namespaces.Get(id)
+		if !ok {
+			return bootstrap.NamespaceResults{},
+				fmt.Errorf("namespace prepared request not found: %v", id.String())
+		}
+
+		// Shallow copy the current namespace for the next namespace prepared request.
+		nextNamespace := currNamespace
+
+		// Calculate bootstrap time ranges.
+		dataRequired := namespace.DataRunOptions.ShardTimeRanges.Copy()
+		dataCurrRequested := currNamespace.DataRunOptions.ShardTimeRanges.Copy()
+		dataCurrFulfilled := dataCurrRequested.Copy()
+		dataCurrFulfilled.Subtract(currResult.DataResult.Unfulfilled())
+
+		dataUnfulfilled := dataRequired.Copy()
+		dataUnfulfilled.Subtract(dataCurrFulfilled)
+
+		// Modify the unfulfilled result.
+		currResult.DataResult.SetUnfulfilled(dataUnfulfilled.Copy())
+
+		// Set the next bootstrapper required ranges.
+		nextNamespace.DataRunOptions.ShardTimeRanges = dataUnfulfilled.Copy()
+
+		var (
+			indexRequired      = result.ShardTimeRanges{}
+			indexCurrRequested = result.ShardTimeRanges{}
+			indexCurrFulfilled = result.ShardTimeRanges{}
+		)
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			// Calculate bootstrap time ranges.
+			indexRequired = namespace.IndexRunOptions.ShardTimeRanges.Copy()
+			indexCurrRequested = currNamespace.IndexRunOptions.ShardTimeRanges.Copy()
+			indexCurrFulfilled = indexCurrRequested.Copy()
```

Comment:
nit: don't think this copy is necessary

Reply:
True, I'm going to opt to just copy everywhere though if that's ok for this code path since it only executes less than 3 times each bootstrap, and easier to not think and just copy it at every single place for safety. (I'm worried about preconditions/postconditions changing and/or others updating this and not copying then something funky going on)